runtime/tchannel_server.go

```diff
@@ -64,197 +84,364 @@ func (ncr netContextRouter) Handle(ctx netContext.Context, call *tchannel.Inboun
+	var err error
+	errc := make(chan error, 1)
+	c := tchannelInboundCall{
+		endpoint: e,
+		call:     call,
 	}
-}
-
-func (s *TChannelRouter) onError(err error) {
-	if tchannel.GetSystemErrorCode(err) == tchannel.ErrCodeTimeout {
-		s.logger.Warn("Thrift server timeout",
-			zap.String("error", err.Error()),
-		)
-	} else {
-		s.logger.Error("Thrift server error.",
-			zap.String("error", err.Error()),
-		)
+	c.start()
+	go func() { errc <- s.handle(ctx, e, &c) }()
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+		if err == context.Canceled {
+			// check if context was Canceled due to handle response
+			if c.responded {
+				err = <-errc
+			}
+		}
+	case err = <-errc:
 	}
+	c.finish(err)
 }
 
-func (s *TChannelRouter) handle(ctx context.Context, handler handler, service string, method string, call *tchannel.InboundCall) error {
-	treader, err := call.Arg2Reader()
+func (s *TChannelRouter) handle(
+	ctx context.Context,
+	e *TChannelEndpoint,
+	c *tchannelInboundCall,
+) (err error) {
+	// read request
+	if err = c.readReqHeaders(ctx); err != nil {
+		return err
+	}
+	wireValue, err := c.readReqBody(ctx)
 	if err != nil {
-		return errors.Wrapf(err, "could not create arg2reader for inbound call: %s::%s", service, method)
+		return err
 	}
 
-	headers, err := ReadHeaders(treader)
-	if err != nil {
-		_ = treader.Close()
+	// trace request
+	tracer := tchannel.TracerFromRegistrar(s.registrar)
+	ctx = tchannel.ExtractInboundSpan(ctx, c.call, c.reqHeaders, tracer)
 
-		return errors.Wrapf(err, "could not reade headers for inbound call: %s::%s", service, method)
+	// handle request
+	resp, err := c.handle(ctx, &wireValue)
+	if err != nil {
+		return err
 	}
-	if err := EnsureEmpty(treader, "reading request headers"); err != nil {
-		_ = treader.Close()
 
-		return errors.Wrapf(err, "could not ensure arg2reader is empty for inbound call: %s::%s", service, method)
+	// write response
+	if err = c.writeResHeaders(ctx); err != nil {
+		return err
 	}
-
-	if err := treader.Close(); err != nil {
-		return errors.Wrapf(err, "could not close arg2reader for inbound call: %s::%s", service, method)
+	if err = c.writeResBody(ctx, resp); err != nil {
+		return err
 	}
 
-	treader, err = call.Arg3Reader()
+	return err
+}
+
+func (c *tchannelInboundCall) start() {
+	c.startTime = time.Now()
+}
+
+func (c *tchannelInboundCall) finish(err error) {
+	c.finishTime = time.Now()
+
+	// emit metrics
 	if err != nil {
-		return errors.Wrapf(err, "could not create arg3reader for inbound call: %s::%s", service, method)
+		c.endpoint.metrics.SystemErrors.Inc(1)
+	} else if !c.success {
+		c.endpoint.metrics.AppErrors.Inc(1)
+	} else {
+		c.endpoint.metrics.Success.Inc(1)
 	}
+	c.endpoint.metrics.Latency.Record(c.finishTime.Sub(c.startTime))
 
-	buf := GetBuffer()
-	defer PutBuffer(buf)
-	if _, err := buf.ReadFrom(treader); err != nil {
-		_ = treader.Close()
+	// write logs
+	OnError(c.endpoint.Logger, err, "Thrift server timeout", "Thrift server error")
```

Comment:
nitpick: longer names

`Got an error whilst finishing a server TChannel request `

We can use the same message for both warn / error and then use the fields on the log to filter out timeouts.

Reply:
I wasn't sure if you wanted to maintain parity with https://github.com/uber/tchannel-go/blob/dev/thrift/server.go#L97, but I also prefer having the same msg for both warn and error.  We don't need the `onError()` function then, `logErrors` will suffice.